### PR TITLE
Theme Support

### PR DESCRIPTION
**Pull request by domoritz**
Replaces https://github.com/vega/editor/pull/360

### CODE REVIEW

File: src/components/input-panel/compiled-spec-header/index.tsx [hunk 1]
```diff
@@ -30,18 +33,51 @@ class CompiledSpecDisplayHeader extends React.Component<Props> {
...
+              className={this.props.sidePaneItem === SIDEPANE.Theme ? 'active-tab' : undefined}
+              onClick={e => {
+                if (this.props.sidePaneItem === SIDEPANE.Theme) {
+                  e.stopPropagation();
+                }
+                e.stopPropagation();
+                this.props.setSidePaneItem(SIDEPANE.Theme);
+              }}
+            >
+              Theme Settings
```
Comment (domoritz) [inline src/components/input-panel/compiled-spec-header/index.tsx]: A theme is an instance of a config. Therefore, better call this "Config".

File: src/components/theme-editor/theme-editor-header.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import { ChevronDown, ChevronUp } from 'react-feather';
+import { connect } from 'react-redux';
+import { bindActionCreators, Dispatch } from 'redux';
+import * as themes from 'vega-themes';
+import * as EditorActions from '../../actions/editor';
+import './theme-editor.css';
+
+class ThemeEditorHeader extends React.Component<any, any> {
```
Comment (domoritz) [inline src/components/theme-editor/theme-editor-header.tsx]: Rename to ConfigEditorHeader

Comment (domoritz) [reply src/components/theme-editor/theme-editor-header.tsx]: And also rename all other mentions of theme to config.

File: src/components/theme-editor/theme-editor-header.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  public render() {
+    return (
+      <div className="theme-header" style={{ position: 'static' }}>
+        <select
```
Comment (domoritz) [inline src/components/theme-editor/theme-editor-header.tsx]: Add a description. `Theme: ` or something.

File: src/components/theme-editor/theme-editor-header.tsx [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        >
+          <option value="custom">Custom</option>
+          {Object.keys(themes).map((keyName, keyIndex) => {
+            return (
+              <option key={keyName} value={keyName}>
+                {keyName}
+              </option>
+            );
+          })}
+        </select>
```
Comment (domoritz) [inline src/components/theme-editor/theme-editor-header.tsx]: Let's try to use react select here.

File: src/components/theme-editor/index.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+class ThemeEditor extends React.Component<any, any> {
+  public handleEditorChange = spec => {
+    this.props.setCurrentTheme(JSON.parse(spec));
+    (document.getElementById('theme_select') as any).value = 'custom';
+  };
+  public render() {
+    return (
+      <div className="sizeFixEditorParent full-height-wrapper">
+        {/* <ThemeEditorHeader arrowToggle={this.props.arrowToggle} /> */}
```
Comment (domoritz) [inline src/components/theme-editor/index.tsx]: What's this?

File: src/components/input-panel/compiled-spec-header/index.tsx [hunk 6]
```diff
@@ -30,18 +33,51 @@ class CompiledSpecDisplayHeader extends React.Component<Props> {
...
+              <ThemeEditorHeader />
+            </>
+          )}
         </div>
       );
     } else {
       return (
         <div onClick={this.props.toggleCompiledVegaSpec} className="editor-header" style={toggleStyle}>
-          <span>Compiled Vega</span>
+          <span>{this.props.mode === Mode.VegaLite ? 'Compiled Vega And' : null} Theme Support</span>
```
Comment (domoritz) [inline src/components/input-panel/compiled-spec-header/index.tsx]: Do not capitalize And.

**Comment by siddhant1** [comment]
Cool! I will have a look

**Comment by siddhant1** [comment]
Made some requested changes, for the custom problem I will keep that in the redux store and generate a fix by tomorrow because its a big refactor

File: src/components/theme-editor/index.tsx [hunk 7]
```diff
@@ -5,18 +5,16 @@ import { connect } from 'react-redux';
 import { bindActionCreators, Dispatch } from 'redux';
 import { debounce } from 'vega';
 import * as EditorActions from '../../actions/editor';
-import ThemeEditorHeader from './theme-editor-header';
 import './theme-editor.css';
```
Comment (domoritz) [inline src/components/theme-editor/index.tsx]: Files are still called `theme`

**Comment by domoritz** [comment]
There are still a number of things left. I marked this PR as wip

**Comment by siddhant1** [comment]
Cool! Generating fixes

**Comment by siddhant1** [comment]
@domoritz All of the issues are fixed (react-select remaining)

File: src/components/input-panel/compiled-spec-header/index.tsx (hunk 1 above)
Comment (domoritz) [reply src/components/input-panel/compiled-spec-header/index.tsx]: Still todo, no?

File: src/components/renderer/renderer.tsx [hunk 8]
```diff
@@ -59,7 +60,7 @@ class Editor extends React.Component<Props, State> {
   }
   // Initialize the view instance
   public initView() {
-    const runtime = vega.parse(this.props.vegaSpec);
+    const runtime = vega.parse(this.props.vegaSpec, this.props.themeConfig);
```
Comment (domoritz) [inline src/components/renderer/renderer.tsx]: just call it config

File: src/reducers/index.ts [hunk 9]
```diff
@@ -276,6 +279,21 @@ export default (state: State = DEFAULT_STATE, action: Action): State => {
         ...state,
         navItem: action.navItem,
       };
+    case SET_CURRENT_THEME:
```
Comment (domoritz) [inline src/reducers/index.ts]: Why no SET_CURRENT_CONFIG?

File: src/actions/editor.ts [hunk 10]
```diff
@@ -53,7 +53,7 @@ export type Action =
   | SetDebugPaneSize
   | ShowLogs
   | SetCompiledVegaPaneSize
-  | SetCurrentTheme
+  | SetCurrentConfig
```
Comment (domoritz) [inline src/actions/editor.ts]: Why not just SetConfig and SetThemeName?

Comment (siddhant1) [reply src/actions/editor.ts]: Will do!

**Comment by siddhant1** [comment]
Fixed! @domoritz please verify

File: src/actions/editor.ts [hunk 11]
```diff
@@ -53,7 +53,7 @@ export type Action =
   | SetDebugPaneSize
   | ShowLogs
   | SetCompiledVegaPaneSize
-  | SetCurrentConfig
+  | SetConfig
   | SetCurrentThemeName
```
Comment (domoritz) [inline src/actions/editor.ts]: Why not `SetThemeName` only?

**Comment by domoritz** [comment]
Hmm, the down icon is still missing.

**Comment by siddhant1** [comment]
I know ! I removed that because it was overlapping over the config tab. I can add it again and make a pr! Maybe you can check if it looks good.

**Comment by domoritz** [comment]
I added it back already.

**Pull request merged**